### ISSUE TITLE
during shutdown, when in the final killing spree a process doesn't react to SIGTERM log about it quickly, even though we SIGKILL much later

### ISSUE THREAD

**Issue by thelounge-zz** [systemd/systemd#11821]
systemd-241-2.gita09c170.fc30.x86_64

Reached target Unmount All Filesystems
Stopped target Swap
Reached Target Shutdown
Reached Target Final Step
Started Reboot
systemd-shutdown: Syncing filesysetms and block devides
systemd-shutdown: Sending SIGTERM to remeining processes
systemd-journald: Received SIGTERM from PID 1 (systemd-shutdown)

and than the VM hangs for a minute or so - for days on this testing VM i thought shutdown don't work at all and hard-reset the guest until i realized now "hey, after ages it recovers itself" while i really don't get what to wait for at this final point of shutdown

hopefully that get fixed before Fedora 30 goes GA given that normally a complete reboot takes 5 seconds at all on such a stripped down vmware guest

**Comment by poettering** [comment]
Hmm, systemd-shutdown is actually invoked at the very very end of things, it doesn't do much, except transition back to the initrd, which then does the rest and reboots.

Please following this guideline:

https://freedesktop.org/wiki/Software/systemd/Debugging/#index2h1

To figure out what is going on there. My educated guess is though that this is already outside of systemd's control and inside dracut/kernel

**Comment by thelounge-zz** [comment]
besdies that i don't get why there is all that "hand back to initrd" after unmount everything is needed instead just reboot/poweroff and "reboot -f" just works so the kernel likely is not the problem see end of that large log, i guess the long pause is the interesting part

[   66.434571] systemd[1]: local-fs-pre.target changed active -> dead
[   66.434578] systemd[1]: local-fs-pre.target: Job 486 local-fs-pre.target/stop finished, result=done
[   66.434580] systemd[1]: Stopped target Local File Systems (Pre).
[   66.436158] systemd[1]: systemd-tmpfiles-setup-dev.service: Succeeded.
[   66.436162] systemd[1]: systemd-tmpfiles-setup-dev.service: Changed exited -> dead
[   66.436176] systemd[1]: systemd-tmpfiles-setup-dev.service: Job 467 systemd-tmpfiles-setup-dev.service/stop finished, result=done
[   66.436178] systemd[1]: Stopped Create Static Device Nodes in /dev.
[   66.436600] systemd[1]: systemd-remount-fs.service: Succeeded.
[   66.436603] systemd[1]: systemd-remount-fs.service: Changed exited -> dead
[   66.436616] systemd[1]: systemd-remount-fs.service: Job 480 systemd-remount-fs.service/stop finished, result=done
[   66.436618] systemd[1]: Stopped Remount Root and Kernel File Systems.
[   66.436924] systemd-journald[370]: Successfully sent stream file descriptor to service manager.
[   66.437409] systemd[1]: systemd-fsck-root.service: Succeeded.
[   66.437413] systemd[1]: systemd-fsck-root.service: Changed exited -> dead
[   66.437787] systemd[557]: tmp.mount: Executing: /usr/bin/umount /tmp -c
[   66.438822] systemd[1]: systemd-fsck-root.service: Job 527 systemd-fsck-root.service/stop finished, result=done
[   66.438826] systemd[1]: Stopped File System Check on Root Device.
[   66.439028] systemd[1]: system-systemd\x2dfsck.slice changed active -> dead
[   66.439327] systemd[1]: system-systemd\x2dfsck.slice: Job 532 system-systemd\x2dfsck.slice/stop finished, result=done
[   66.439330] systemd[1]: Removed slice system-systemd\x2dfsck.slice.
[   66.440346] systemd[1]: systemd-journald.service: Got notification message from PID 370 (FDSTORE=1)
[   66.440359] systemd[1]: systemd-journald.service: Added fd 26 (n/a) to fd store.
[   66.440897] systemd[1]: libmount event [rescan: yes]
[   66.442015] systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 26 (stored), closing.
[   66.442265] systemd[1]: Received SIGCHLD from PID 557 (umount).
[   66.442290] systemd[1]: Child 557 (umount) died (code=exited, status=0/SUCCESS)
[   66.442315] systemd[1]: tmp.mount: Child 557 belongs to tmp.mount.
[   66.442325] systemd[1]: tmp.mount: Mount process exited, code=exited, status=0/SUCCESS
[   66.442328] systemd[1]: tmp.mount: Succeeded.
[   66.442333] systemd[1]: tmp.mount: Changed unmounting -> dead
[   66.443038] systemd[1]: tmp.mount: Job 545 tmp.mount/stop finished, result=done
[   66.443042] systemd[1]: Unmounted Temporary Directory (/tmp).
[   66.443322] systemd[1]: umount.target changed dead -> active
[   66.443329] systemd[1]: umount.target: Job 542 umount.target/start finished, result=done
[   66.443331] systemd[1]: Reached target Unmount All Filesystems.
[   66.443504] systemd[1]: swap.target changed active -> dead
[   66.443510] systemd[1]: swap.target: Job 476 swap.target/stop finished, result=done
[   66.443513] systemd[1]: Stopped target Swap.
[   66.443676] systemd[1]: shutdown.target changed dead -> active
[   66.443682] systemd[1]: shutdown.target: Job 453 shutdown.target/start finished, result=done
[   66.443684] systemd[1]: Reached target Shutdown.
[   66.443846] systemd[1]: final.target changed dead -> active
[   66.443852] systemd[1]: final.target: Job 541 final.target/start finished, result=done
[   66.443854] systemd[1]: Reached target Final Step.
[   66.444522] systemd[1]: systemd-reboot.service: Changed dead -> start
[   66.444571] systemd[1]: systemd-reboot.service: Succeeded.
[   66.444574] systemd[1]: systemd-reboot.service: Changed start -> dead
[   66.444620] systemd[1]: systemd-reboot.service: Job 452 systemd-reboot.service/start finished, result=done
[   66.444622] systemd[1]: Started Reboot.
[   66.444774] systemd[1]: Forcibly rebooting: unit systemd-reboot.service succeeded
[   66.444792] systemd[1]: reboot.target changed dead -> active
[   66.444798] systemd[1]: reboot.target: Job 451 reboot.target/start finished, result=done
[   66.444800] systemd[1]: Reached target Reboot.
[   66.444951] systemd[1]: Shutting down.
[   66.447773] systemd[1]: systemd-journald.service: Releasing resources.
[   66.447776] systemd[1]: systemd-journald.service: Releasing all stored fds
[   66.465137] systemd-shutdown[1]: Found cgroup2 on /sys/fs/cgroup/unified, unified hierarchy for systemd controller
[   66.473835] systemd-shutdown[1]: Successfully forked off '(sd-sync)' as PID 558.
[   66.473841] systemd-shutdown[1]: Syncing filesystems and block devices.
[   66.475408] systemd-shutdown[1]: Sending SIGTERM to remaining processes...
[   66.477127] systemd-journald[370]: Received SIGTERM from PID 1 (systemd-shutdow).
[   66.477140] systemd-journald[370]: systemd-journald stopped as pid 370
[  156.479232] systemd-shutdown[1]: Sending SIGKILL to remaining processes...
[  156.482141] systemd-shutdown[1]: Sending SIGKILL to PID 374 (rngd).
[  156.486048] systemd-shutdown[1]: Unmounting file systems.
[  156.486960] systemd-shutdown[1]: Successfully forked off '(sd-remount)' as PID 559.
[  156.487272] [559]: Remounting '/' read-only in with options 'commit=30,inode_readahead_blks=16'.
[  156.490156] EXT4-fs (sdb1): re-mounted. Opts: commit=30,inode_readahead_blks=16
[  156.491622] systemd-shutdown[1]: All filesystems unmounted.
[  156.491626] systemd-shutdown[1]: Deactivating swaps.
[  156.491696] systemd-shutdown[1]: All swaps deactivated.
[  156.491699] systemd-shutdown[1]: Detaching loop devices.
[  156.491714] systemd-shutdown[1]: sd-device-enumerator: Scan all dirs
[  156.491741] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/bus
[  156.491774] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/class
[  156.491826] systemd-shutdown[1]: All loop devices detached.
[  156.491828] systemd-shutdown[1]: Detaching DM devices.
[  156.491833] systemd-shutdown[1]: sd-device-enumerator: Scan all dirs
[  156.491844] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/bus
[  156.491862] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/class
[  156.491892] systemd-shutdown[1]: All DM devices detached.
[  156.491896] systemd-shutdown[1]: All filesystems, swaps, loop devices and DM devices detached.
[  156.492093] systemd-shutdown[1]: Successfully forked off '(sd-executor)' as PID 560.
[  156.495550] [560]: Successfully forked off '(direxec)' as PID 561.
[  156.500971] EXT4-fs (sdb1): re-mounted. Opts: commit=30,inode_readahead_blks=16

**Comment by benjarobin** [comment]
I can reproduce this issue, running on Arch Linux with:
 - kernel 5.0.4
 - systemd 241.7-2 (The Arch Linux version of systemd), the commit is :  a09c170122cf3b37c3e4431bf082f9dbdc07fc70
 - Or with systemd from master (fd7887da006a0d64dc4e1eeb6c0c9ef197a5e500)

The last line seen from the console is `Reached target Power-Off`, and if I wait 1 minute, the PC power off.

With systemd 240 (commit 0430af90ef28a229e51136ad364e1da7e76c8638) I do not have any problem.

But I have a lot of trouble with the debugging:
 - This is kind of random, reproduce rate of 70% (approximately)
 - I did not reproduce with kernel 4.19.30 but I did not do a lot of testing with this kernel version
 - Cannot use debug-shell.service at this point: I can switch to the debug shell and I can see the prompt, but the shell is no longer "alive"
 - Adding one of these kernel parameter "fix" the issue (so adding more log or showing them on the console is not going to help):
    - `systemd.log_level=debug`
    - `systemd.log_target=console`

**Comment by benjarobin** [comment]
If I follow https://freedesktop.org/wiki/Software/systemd/Debugging/#index3h2 I can reproduce the issue, here the end of the `shutdown-log.txt` file:
```
[  128.233235] systemd[1]: Reached target Power-Off.
[  128.233278] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.233345] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.234102] systemd[1]: lvm2-monitor.service: About to execute: /usr/bin/lvm vgchange --monitor n --ignoreskippedcluster
[  128.234285] systemd[1]: lvm2-monitor.service: Forked /usr/bin/lvm as 1419
[  128.234395] systemd[1]: lvm2-monitor.service: Changed exited -> stop
[  128.234404] systemd[1]: Stopping Monitoring of LVM2 mirrors, snapshots etc. using dmeventd or progress polling...
[  128.235052] systemd[1]: Shutting down.
[  128.235303] systemd-journald[381]: Successfully sent stream file descriptor to service manager.
[  128.235417] systemd[1419]: lvm2-monitor.service: Executing: /usr/bin/lvm vgchange --monitor n --ignoreskippedcluster
[  128.235705] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235761] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235791] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235817] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235857] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235883] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235907] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235933] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235958] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.235998] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.237144] systemd[1]: systemd-journald.service: Releasing resources.
[  128.237147] systemd[1]: systemd-journald.service: Releasing all stored fds
[  128.238986] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.239024] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.247342] systemd-shutdown[1]: Found cgroup2 on /sys/fs/cgroup/unified, unified hierarchy for systemd controller
[  128.249059] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.249792] systemd-shutdown[1]: Successfully forked off '(sd-sync)' as PID 1423.
[  128.249795] systemd-shutdown[1]: Syncing filesystems and block devices.
[  128.252357] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.252481] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.271780] systemd-shutdown[1]: Sending SIGTERM to remaining processes...
[  128.277069] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  128.277192] systemd-journald[381]: Received SIGTERM from PID 1 (systemd-shutdow).
[  128.277198] systemd-journald[381]: systemd-journald stopped as pid 381
[  128.277208] systemd-journald[381]: Journal effective settings seal=no compress=yes compress_threshold_bytes=512B
[  218.267027] systemd-shutdown[1]: Sending SIGKILL to remaining processes...
[  218.270501] systemd-shutdown[1]: Sending SIGKILL to PID 386 (lvmetad).
[  218.272345] systemd-shutdown[1]: Unmounting file systems.
[  218.273106] systemd-shutdown[1]: Successfully forked off '(sd-remount)' as PID 1426.
[  218.273487] [1426]: Remounting '/' read-only in with options '(null)'.
[  218.313286] EXT4-fs (sda4): re-mounted. Opts: (null)
[  218.337377] systemd-shutdown[1]: All filesystems unmounted.
[  218.337379] systemd-shutdown[1]: Deactivating swaps.
[  218.337423] systemd-shutdown[1]: All swaps deactivated.
[  218.337424] systemd-shutdown[1]: Detaching loop devices.
[  218.337434] systemd-shutdown[1]: sd-device-enumerator: Scan all dirs
[  218.337460] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/bus
[  218.337487] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/class
[  218.337532] systemd-shutdown[1]: All loop devices detached.
[  218.337533] systemd-shutdown[1]: Detaching DM devices.
[  218.337535] systemd-shutdown[1]: sd-device-enumerator: Scan all dirs
[  218.337540] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/bus
[  218.337549] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/class
[  218.337567] systemd-shutdown[1]: All DM devices detached.
[  218.337569] systemd-shutdown[1]: All filesystems, swaps, loop devices and DM devices detached.
[  218.337716] systemd-shutdown[1]: Successfully forked off '(sd-executor)' as PID 1427.
[  218.340689] [1427]: Successfully forked off '(direxec)' as PID 1428.
[  218.340987] [1427]: Successfully forked off '(direxec)' as PID 1429.
[  218.350871] EXT4-fs (sda4): re-mounted. Opts: data=ordered
[  218.353085] [1427]: /usr/lib/systemd/system-shutdown/mdadm.shutdown succeeded.
```

**Comment by benjarobin** [comment]
I create the following patch in order to debug it, It can be improved, and I will create a MR with it:
```
diff --git a/src/core/killall.c b/src/core/killall.c
index f0ce996556..d5f25b73ac 100644
--- a/src/core/killall.c
+++ b/src/core/killall.c
@@ -133,8 +133,18 @@ static void wait_for_children(Set *pids, sigset_t *mask, usec_t timeout) {
                         return;
 
                 n = now(CLOCK_MONOTONIC);
-                if (n >= until)
+                if (n >= until) {
+                        log_warning("Children left (not killed):");
+
+                        SET_FOREACH(p, pids, i) {
+                                _cleanup_free_ char *cmdline = NULL;
+
+                                if (get_process_cmdline(PTR_TO_PID(p), 80, true, &cmdline) >= 0) {
+                                        log_warning("  %s", cmdline);
+                                }
+                        }
                         return;
+                }
 
                 timespec_store(&ts, until - n);
                 k = sigtimedwait(mask, NULL, &ts);
```
This show the following lines:
```
[ 2665.318703] systemd-shutdown[1]: Children left (not killed):
[ 2665.318743] systemd-shutdown[1]:   /usr/bin/lvmetad -f
```

**Comment by benjarobin** [comment]
In the previous boot log (using journalctl or in `shutdown-log.tx`), I never see the following lines with systemd 241 (nor with systemd 240), why ?
```
Stopping LVM2 metadata daemon...
Stopped LVM2 metadata daemon.
```

**Comment by benjarobin** [comment]
In order to fix it, I just add the following line to `lvm2-lvmetad.service` and `lvm2-lvmetad.socket`:
```
Before=shutdown.target
```

@poettering  Is it the proper fix ? If yes I should report to the LVM2 project...

**Comment by arvidjaar** [comment]
> I should report to the LVM2 project

`lvmetad` no more exists in LVM2.

**Comment by arvidjaar** [comment]
> I create the following patch in order to debug it

was not

> `[  218.270501] systemd-shutdown[1]: Sending SIGKILL to PID 386 (lvmetad).`

already providing the same information?

**Comment by benjarobin** [comment]
@arvidjaar Indeed the MR is not needed at all. How can I missed that, what a wast of time... 

And for LVM2 you are right, in master `lvmetad` was removed, but it is still present in the latest release `v2_02_184`

I just want to know if my fix is right, if this is the case I will create a bug report on Arch Linux in order to add a patch on the Arch package

**Comment by arvidjaar** [comment]
> I just want to know if my fix is right
> This imply that the problem reported by @thelounge-zz is not the same

Actually it is the same problem. There is race condition between stopping units and systemd getting request to exit main loop, so it proceeds with shutdown before all pending stop jobs have been processed.

If in the past systemd guaranteed that all pending jobs are processed before jumping to final shutdown, it is regression. If this happened to work as a side effect, requirement to order units before `shutdown.target` (or at least `final.target`) needs to be clearly communicated.

**Comment by benjarobin** [comment]
I am not agree with you, in my case the service (https://sourceware.org/git/?p=lvm2.git;a=blob;f=scripts/lvm2_lvmetad_systemd_red_hat.service.in;h=92e6d695f157a28739f37a1afafdd2f471466a8e;hb=7cbee7e9cf5445ee5c50898e20536734cbb0f4a6) clearly specify to ignore default dependency, so the service (if I understand things correctly) is not stopped. The only thing that I do not understand, is why the SIGTERM signal did not stopped the service

**Comment by arvidjaar** [comment]
> clearly specify to ignore default dependency

It also clearly specifies `Conflicts=shutdown.target`

> so the service is not stopped

It is **queued** to be stopped. The fact that it is not stopped is a bug, not a feature.

**Comment by benjarobin** [comment]
No this is not a bug, this is specified in the documentation https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Conflicts=

If you want that the unit to be stopped before the specified target, you need to specify both Conflicts= and Before=

**Comment by patrakov** [comment]
I was affected by the same bug. Adding `Before=shutdown.target` via `systemctl edit lvm2-lvmetad` helped.

**Comment by thelounge-zz** [comment]
here runs a minimal setup with no lvm/raid and so on - soemtimes it reboots/shutdowns as excepected, sometimes it hangs around for ages for no good reason

**Comment by benjarobin** [comment]
In your case, this is rngd the problem. You should check how the service is started, and provide here the content of the service. You can also try to stop manually this service before trying to shutdown your PC

**Comment by thelounge-zz** [comment]
look at the time stamp, when this setting exists for a full year (in fact it exists longer and i switched from a complete override to a dropin a year ago) and suddenly systemd becomes too stupid to handle it correctly don't explain me where the regression is

[root@rawhide ~]# cat /etc/systemd/system/rngd.service.d/security.conf
[Service]
ReadOnlyPaths=/etc
ReadOnlyPaths=/usr
SystemCallFilter=~@clock @cpu-emulation @debug @module @mount @reboot @swap

[root@rawhide ~]# ls /etc/systemd/system/rngd.service.d/default-dependencies.conf
-rw-r--r-- 1 root root 30 2018-05-27 04:06 /etc/systemd/system/rngd.service.d/default-dependencies.conf

[root@rawhide ~]# cat /etc/systemd/system/rngd.service.d/default-dependencies.conf
[Unit]
DefaultDependencies=no

**Comment by poettering** [comment]
@thelounge-zz it appears rngd doesn't react to SIGTERM on your system and thus we resort to using SIGKILL after a time-out. That's the delay you are seeing.

And yes in @benjarobin's case it's that LVM daemon that refuses to shutting down after SIGTERM, and we thus eventually have to SIGKILL.

I don#t think there's anything to fix on systemd's side here. The log output already makes clear what is being SIGKILLed because it didn't shut down, hence systemd is actually able to recover from this correctly. The right fixes are to find out why rngd and that LVM daemon decide to hang and refuse shutdown on SIGTERM, but that's not something we can determine from systemd's side.

Anyway, I am inclined to close this here. Or am I missing something?

**Comment by thelounge-zz** [comment]
yes! you miss that this late at shutdown when everything is alrteady unmounted there is no need to wait that long at all

**Comment by patrakov** [comment]
@benjarobin in the LVM case it's pretty clear. lvmetad refuses to react to SIGTERM if there are any volumes that need monitoring. Both lvm2-lvmetad.service and lvm2-monitor.service units mistakenly specify `Conflicts=shutdown.target`, so there is nothing that queues them before that target. I actually have a screenshot that clearly shows activity related to stopping LVM2 monitoring after the shutdown.target has been reached.

So we just need to make sure that monitoring is stopped, and then lvmetad is stopped with SIGTERM, before shutdown. Note that lvm2-monitor.service already has `After=... lvm2-lvmetad.socket lvm2-lvmetad.service`, so lvmetad will be stopped (with SIGTERM) after stopping the monitoring. So the whole problem is that nothing orders stopping lvmetad before shutdown.target. I.e. the whole fix is:

```
# /etc/systemd/system/lvm2-lvmetad.service.d/override.conf
[Unit]
Before=shutdown.target
```

**Comment by patrakov** [comment]
@thelounge-zz false, if you don't wait for LVM or RAID monitoring to stop, then you will lose data. It's just that we didn't convince lvmetad (by stopping monitoring) that it is safe to stop on SIGTERM. Indeed for some daemons there is nothing to lose due to early SIGKILL, but there is no way for systemd to distinguish daemons that are safe to kill quickly from those that aren't.

**Comment by benjarobin** [comment]
@thelounge-zz In your case, the problem comes from the following added file : `/etc/systemd/system/rngd.service.d/default-dependencies.conf`

**Comment by thelounge-zz** [comment]
i strongly doubt that the kernel needs to have mdadm hanging around at all to have a raid working and write it#s data

**Comment by patrakov** [comment]
@thelounge-zz for the normal case with standard metadata, it indeed doesn't have mdmon to hang around. But with firmware RAID, due to non-standard metadata, it punts the task of altering the dirty bitmaps to userspace - i.e. mdmon.

**Comment by thelounge-zz** [comment]
one last time: DefaultDependencies=no is set for years on dozens of machines for rngd.service, period

**Comment by poettering** [comment]
> yes! you miss that this late at shutdown when everything is alrteady unmounted there is no need to wait that long at all - you are done at that point, just trigger reboot/shutdown

but how would we know that? I mean, we don't know if rngd is important or not. It could be some storage daemon that syncs stuff to disk and thus takes a bit. I mean, yes, we need to kill it eventually, but we I don't think there's much value in discussing whether it should be 5s or 60s. The fact is that a service neither should have survived that long nor refuse SIGTERM, and this is just the third safety net in effect. I think it's entirely fine if the third safety net is enforcing a long timeout, because it should never be triggered in clean code paths anyway... Or to say this differently: it's preferable to fix/focus on the fact that the service survived that long into shutdown and refuses SIGTERM...

**Comment by poettering** [comment]
> one last time: DefaultDependencies=no is set for years on dozens of machines for rngd.service, period

So if things changed then for some reason rngd.service apparently started to ignore SIGTERM recently, and didn't use to?

**Comment by poettering** [comment]
> > one last time: DefaultDependencies=no is set for years on dozens of machines for rngd.service, period
> 
> So if things changed then for some reason rngd.service apparently started to ignore SIGTERM recently, and didn't use to?

Also, why do you set DefaultDependencies=no like that? if you want to have it in earliest boot, then that's fine, but you could still order it properly on shutdown.target, because you shouldn't need it then...

**Comment by thelounge-zz** [comment]
than frankly give some useful output instead a dumb "Shutting down." after "OK Reached taregt Reboot" which let the machine eppear to be frozen and never come back 

that leaded to some hard resets by me because nothing told me "i wait uelsess for something" or even for what it is waiting - that's that many years afetr systemd was introduced a poor user expierience and again it's hard to argue against systemd haters when they ciome up with such things and print out that was never a problem with sysvinit when you try to shut them down with arguments

**Comment by patrakov** [comment]
Apparently we are talking about two units that made the same mistake:

```
DefaultDependencies=no
Conflicts=shutdown.target
# no Before=shutdown.target
```

(or the same with umount.target)

Probably we need writing a linter and scheduling a mass check of all units in, say, Fedora.

**Comment by poettering** [comment]
converted this into an RFE bug then asking for displaying which process is refusing to exit after SIGTERM

**Comment by topimiettinen** [comment]
@patrakov my Debian Buster system:
```
# find /usr/lib/systemd/system/*.service -type f | xargs grep --files-with-matches DefaultDependencies=no | xargs grep --files-with-matches Conflicts=shutdown.target | wc -l
38
```

**Comment by mbiebl** [comment]
@topimiettinen We could turn that into a lintian test. Should be rather simple to do

**Comment by mbiebl** [comment]
Picking up the suggestions from https://github.com/systemd/systemd/issues/11821#issuecomment-477553923

**Comment by thelounge-zz** [comment]
well, i had these options in my rngd-dropin and one out of 15 reboots still did hang 1.5 minutes at "wating for service to stop..." while this time actually say something useful

DefaultDependencies=no
Conflicts=shutdown.target
Before=shutdown.target

**Comment by topimiettinen** [comment]
@mbiebl I'll try lintian.

**Comment by topimiettinen** [comment]
Actually, with more careful searching, there are not so many matches after all:
```
# find /usr/lib/systemd/system/*.service -type f | xargs grep --files-with-matches DefaultDependencies=no | xargs grep --files-with-matches Conflicts=.*shutdown.target | xargs grep --files-without-match Before=.*shutdown.target
/usr/lib/systemd/system/blk-availability.service
/usr/lib/systemd/system/lvm2-lvmpolld.service
/usr/lib/systemd/system/lvm2-monitor.service
/usr/lib/systemd/system/mdadm-shutdown.service
```
I suppose these are OK. Anyway, I'll submit my new lintian check as it's already done.

**Comment by benjarobin** [comment]
@poettering I was working on the RFE, but I have a slight problem: It looks like the `status_printf()` cannot be used (currently) from the shutdown executable (link problem). What is your recommended way of doing it?

See my current working branch: https://github.com/benjarobin/systemd/tree/killall-show-not-killed

**Comment by topimiettinen** [comment]
@mbiebl Opened Lintian MR [177](https://salsa.debian.org/lintian/lintian/merge_requests/177).

**Comment by benjarobin** [comment]
@poettering  Maybe this issue can be closed since the RFE was implemented in #12153

**Comment by thelounge-zz** [comment]
on Fedora 29 without rngd you seem to have good chances to see that if you used "loginctl enable-linger" for some users, i had seen that in teh past too randomly but not that often, not that long waiting and with some output instead wait silently for nothing - this is *really* a pain in the ass on virtual servers which reboot normally within 5 seconds

**Comment by poettering** [comment]
> @poettering Maybe this issue can be closed since the RFE was implemented in #12153

ok

**Comment by tomachinz** [comment]
For me, it is the process 'sd-sync' and sometimes 'udiskd' that hangs. I will try reboot -f to see if that is faster. Mine always cleanly reboots but takes forever and often tempting to hard reset. It's related to having do 3 a gui login. If I've never logged in at the gui, reboot is very fast.
